notes/fixing-sealing-and-recursive-records.md

```diff
@@ -0,0 +1,227 @@
+# Polymorphic Sealing and Contract Propagation
+
+This draft note was written in the aftermath of us discovering an issue
+involving the interaction of lazy record contracts, recursive records, and
+polymorphic sealing (#1161, #1228).
+
+Nickel implements dynamic contract checks for polymorphic types by using
+a sealing discipline. When converting a static type into a contract, upon
+encountering a universal quantification like `forall a. ...`, Nickel maps `a` to
+a sealing key. Then, the contract checking `a` in positive position becomes an
+application of the `%unseal%` primop with the respective sealing key, while the
+contract checking `a` in negative position becomes an application of `%seal%`.
+For the contract check to ultimately succeed, evaluation must proceed in such a
+way that `%unseal%` ends up only being applied to a term sealed by `%seal%` with
+matching sealing keys. This means that `%seal%` and `%unseal%` must always end
+up being paired one to one in program being evaluated. However, `%seal%` and
+`%unseal%` are generated in different places.
+
+If a contract containing an as yet unpaired `%seal%` or `%unseal%` ends up being
+attached to a record field, any reference to this field will have to apply the
+contract and essentially duplicates the `%seal%` or `%unseal%` operation. This
+means that if a type containing a free type variable is turned into a contract,
+and this contract gets attached to a record field for some reason, any reference
+to the field has the potential to produce a mismatch between `%seal%` and
+`%unseal%` primops in the program.
```

Comment:
Might be useful to reproduce the original example here, just for the sake of (relative) self containment.

Reply:
The example is detailed below, though. Still, it's very drafty, and it's probably better if one can understand what this is about without having to dive into the second part if not needed.